crates/sncast/src/lib.rs

```diff
@@ -220,24 +199,68 @@ pub async fn get_account<'a>(
     keystore: Option<Utf8PathBuf>,
 ) -> Result<SingleOwnerAccount<&'a JsonRpcClient<HttpTransport>, LocalWallet>> {
     let chain_id = get_chain_id(provider).await?;
-    let mut account = if let Some(keystore) = keystore {
-        get_account_from_keystore(provider, chain_id, &keystore, account)?
+    let account_info = if let Some(keystore) = keystore {
+        get_account_data_from_keystore(account, &keystore)?
     } else {
-        get_account_from_accounts_file(account, accounts_file, provider, chain_id)?
+        get_account_data_from_accounts_file(account, chain_id, accounts_file)?
     };
 
-    account.set_block_id(get_block_id("pending")?);
-    verify_account_address(account.clone()).await?;
+    let account = build_account(account_info, chain_id, provider).await?;
+
+    Ok(account)
+}
+
+async fn build_account(
+    account_data: AccountData,
+    chain_id: FieldElement,
+    provider: &JsonRpcClient<HttpTransport>,
+) -> Result<SingleOwnerAccount<&JsonRpcClient<HttpTransport>, LocalWallet>> {
+    let signer = LocalWallet::from(SigningKey::from_secret_scalar(
+        parse_number(&account_data.private_key)
+            .context("Failed to convert private key to FieldElement")?,
+    ));
+    let address = parse_number(&account_data.address).with_context(|| {
+        format!(
+            "Failed to convert account address = {} to FieldElement",
+            &account_data.address
+        )
+    })?;
+    verify_account_address(address, provider).await?;
+
+    let class_hash = account_data
+        .class_hash
+        .map(|class_hash| {
+            parse_number(&class_hash).with_context(|| {
+                format!(
+                    "Failed to convert class hash = {} to FieldElement",
+                    &class_hash
+                )
+            })
+        })
+        .transpose()?;
+
+    let account_encoding =
+        get_account_encoding(account_data.legacy, class_hash, address, provider).await?;
+
+    let mut account =
+        SingleOwnerAccount::new(provider, signer, address, chain_id, account_encoding);
+
+    account.set_block_id(BlockId::Tag(Pending));
 
     Ok(account)
 }
 
-async fn verify_account_address(account: impl ConnectedAccount + std::marker::Sync) -> Result<()> {
-    match account.get_nonce().await {
+async fn verify_account_address(
+    address: FieldElement,
+    provider: &JsonRpcClient<HttpTransport>,
+) -> Result<()> {
+    match provider.get_nonce(BlockId::Tag(Pending), address).await {
         Ok(_) => Ok(()),
         Err(error) => {
             if let StarknetError(ContractNotFound) = error {
-                Err(anyhow!("Invalid account address"))
+                Err(anyhow!(
+                    "No account exists with the provided address {address:#x}"
```

Comment:
nitpick: let's add chain_id to this message as well; just because an address does not exist in network X, it does not mean it doesn't for network Y as well

Reply:
https://github.com/foundry-rs/starknet-foundry/pull/1850/commits/6ac076e32698fd18a88cec63ece2c4766ff9f3ee